controllers/metric_controller.go

```diff
@@ -208,49 +207,145 @@ func (r *MetricReconciler) collectMetrics(ctx context.Context, t *redskyv1beta1.
+func (r *MetricReconciler) target(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) (runtime.Object, error) {
+	if m.Type != redskyv1beta1.MetricKubernetes && m.Type != "" {
+		return nil, nil
+	}
+
+	// Get the target reference, default to the trial itself if there is no reference
+	targetRef := m.Target
+	if m.Target == nil || m.Target.Kind == "" {
+		return t, nil
+	}
 
-		if err = r.List(ctx, target, client.InNamespace(namespace), sel); err != nil {
+	// If no explicit namespace is specified, use the trial namespace
+	if targetRef.Namespace == "" {
+		targetRef.Namespace = t.Namespace
+	}
+
+	// If a name is specified, just get a single object
+	if targetRef.Name != "" {
+		target := &unstructured.Unstructured{}
+		target.SetGroupVersionKind(targetRef.GroupVersionKind())
+		if err := r.Get(ctx, targetRef.NamespacedName(), target); err != nil {
 			return nil, err
 		}
-
 		return target, nil
-	default:
-		// Assume no target is necessary
-		return nil, nil
 	}
+
+	// Convert the selector from a Kubernetes object to something the client can use
+	sel, err := meta.MatchingSelector(targetRef.LabelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch the list of matching resources
+	target := &unstructured.UnstructuredList{}
+	target.SetGroupVersionKind(targetRef.GroupVersionKind())
+	if err := r.List(ctx, target, client.InNamespace(targetRef.Namespace), sel); err != nil {
+		return nil, err
+	}
+	return target, nil
+}
+
+// applyMetricDefaults fills in default values for the supplied metric.
+func (r *MetricReconciler) applyMetricDefaults(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) error {
+	// Give Prometheus metrics a default URL
+	if m.Type == redskyv1beta1.MetricPrometheus && m.URL == "" {
+		m.URL = fmt.Sprintf("http://redsky-%s-prometheus:9090/", t.Namespace)
+	}
+
+	// Default to the trial namespace
+	if m.Target != nil && m.Target.Namespace == "" {
+		// TODO How do we exclude non-namespaced kinds from this default?
+		m.Target.Namespace = t.Namespace
+	}
+
+	// This is strictly for converted v1alpha1 experiments; we should remove it eventually
+	return r.resolveLegacyURL(ctx, t, m)
+}
+
+// resolveLegacyURL checks for the legacy hostname placeholder and replaces it with a hostname determined by
+// looking up a Kubernetes Service object. This roughly corresponds to the original behavior of the controller
+// where URL based metrics were defined using Service selectors instead of actual URLs.
+func (r *MetricReconciler) resolveLegacyURL(ctx context.Context, t *redskyv1beta1.Trial, m *redskyv1beta1.Metric) error {
+	if m.Type != redskyv1beta1.MetricPrometheus && m.Type != redskyv1beta1.MetricJSONPath {
+		return nil
+	}
+
+	// Look for the special placeholder hostname that indicates we should look up a service
+	u, err := url.Parse(m.URL)
+	if err != nil || u.Hostname() != redskyv1alpha1.LegacyHostnamePlaceholder {
```

Comment:
nit, should we split this up?

```
if err != nil {
return err
}

if u.Hostname() != redskyv1alpha1.LegacyHostnamePlaceholder {
return nil
}
```
I know what's here currently works, but this makes it very clear that LegacyHostname is not actually an error condition